c/tskit/core.c

```diff
@@ -1177,3 +1177,97 @@ tsk_avl_tree_int_ordered_nodes(const tsk_avl_tree_int_t *self, tsk_avl_node_int_
     ordered_nodes_traverse(self->head.rlink, 0, out);
     return 0;
 }
+
+/* Bit Array implementation. Allows us to store unsigned integers in a compact manner.
+ * Currently implemented as an array of 32-bit unsigned integers for SIMD friendliness,
+ * (needs more profiling) */
+
+int
+tsk_bit_array_init(tsk_bit_array_t *self, tsk_size_t num_bits, tsk_size_t length)
+{
+    int ret = 0;
+    self->size = (num_bits >> TSK_BIT_ARRAY_CHUNK)
+                 + (num_bits % TSK_BIT_ARRAY_NUM_BITS ? 1 : 0);
+    self->data = tsk_calloc(self->size * length, sizeof(*self->data));
+    if (self->data == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+out:
+    return ret;
+}
+
+void
+tsk_bit_array_get_row(const tsk_bit_array_t *self, size_t row, tsk_bit_array_t *out)
+{
+    out->size = self->size;
+    out->data = self->data + (row * self->size);
+}
+
+void
+tsk_bit_array_intersect(
+    const tsk_bit_array_t *self, const tsk_bit_array_t *other, tsk_bit_array_t *out)
+{
+    for (tsk_size_t i = 0; i < self->size; i++) {
+        out->data[i] = self->data[i] & other->data[i];
+    }
+}
+
+void
+tsk_bit_array_subtract(tsk_bit_array_t *self, const tsk_bit_array_t *other)
+{
+    for (tsk_size_t i = 0; i < self->size; i++) {
+        self->data[i] &= ~(other->data[i]);
+    }
+}
+
+void
+tsk_bit_array_add(tsk_bit_array_t *self, const tsk_bit_array_t *other)
+{
+    for (tsk_size_t i = 0; i < self->size; i++) {
+        self->data[i] |= other->data[i];
+    }
+}
+
+void
+tsk_bit_array_add_bit(tsk_bit_array_t *self, const tsk_bit_array_value_t bit)
+{
+    tsk_bit_array_value_t i = bit >> TSK_BIT_ARRAY_CHUNK;
+    self->data[i] |= (tsk_bit_array_value_t) 1 << (bit - (TSK_BIT_ARRAY_NUM_BITS * i));
+}
+
+bool
+tsk_bit_array_contains(const tsk_bit_array_t *self, const tsk_bit_array_value_t bit)
+{
+    tsk_bit_array_value_t i = bit >> TSK_BIT_ARRAY_CHUNK;
+    return self->data[i]
+           & ((tsk_bit_array_value_t) 1 << (bit - (TSK_BIT_ARRAY_NUM_BITS * i)));
+}
+
+bool
+tsk_bit_array_contains_id(const tsk_bit_array_t *self, const tsk_id_t node)
+{
+    if (node == TSK_NULL) {
+        return false;
+    }
+    return tsk_bit_array_contains(self, (tsk_bit_array_value_t) node);
+}
+
+void
```

Comment:
I've changed this function to return a `tsk_size_t` value, but I've kept the functions that modify a pointer in place, even though they don't raise an error. I think that it's best to avoid copies of the bit arrays.

Reply:
Agreed - I'd just use return values for simple scalars